### ISSUE TITLE
MySQL allows duplicate column names, but Dolt doesn't

### ISSUE THREAD

**Issue by nicktobey** [dolthub/dolt#6676]
Why would you ever want to have duplicate column names in your query results? Beats me, but `sqllogictests` has tests for it. 323 of them, in fact.

In Dolt, attempting to generate a result set with duplicate column names returns an "ambiguous column" error.

In MySQL, the error is deferred until you try to reference an ambiguous column from the result set. This includes:
- Creating a table from the result set
- Using the result set as a subquery (even if the main query doesn't reference the ambiguous columns)
- Presumably, most other operations.

I say "most" because `GROUP BY` and `ORDER BY` are allowed to reference a potentially ambiguous column name, as long as the reference can be unambiguously resolved according to certain rules, which I describe below.

tl;dr: `GROUP BY` is allowed on queries with duplicate column names, provided that the expression for `GROUP BY` is the name of a column on the underlying table. The expression binds to that column. This is allowed even if the name also conflicts with one or more columns in the result set. We need to support this behavior.

Examples:

```sql
CREATE TABLE tab (col int);
INSERT INTO tab VALUES (2), (2), (1);
SELECT col, col FROM tab GROUP BY col;
```

Output:

```
+------+------+
| col  | col  |
+------+------+
|    2 |    2 |
|    1 |    1 |
+------+------+
```

```sql
CREATE TABLE tab (col int);
INSERT INTO tab VALUES (2), (2), (1);
SELECT col, 12 as col from tab group by col;
```

Output:

```
+------+-----+
| col  | col |
+------+-----+
|    2 |  12 |
|    1 |  12 |
+------+-----+
```

**Comment by nicktobey** [comment]
The rest of this comment is the full behavior of `GROUP BY` and `ORDER BY` that I observed. I suspect that a lot of if is the result bugs in MySQL's name resolution. None of this should be needed in order to address the test failures; I only had to figure this out to determine that we *don't* need to match it to satisfy the failing conformance tests. It's here for completeness' sake.

According to the spec:

> MySQL resolves unqualified column or alias references in ORDER BY clauses by searching in the select_expr values, then in the columns of the tables in the FROM clause. For GROUP BY or HAVING clauses, it searches the FROM clause before searching in the select_expr values. 

In practice, the observed behavior differs. I suspect this is a bug.

In practice, `ORDER BY` resolves the column reference by looking for (in order of preference):
- A nontrivial column alias. (By nontrivial, I mean an alias that isn't just a renaming of an underlying column. For example, `col1 as col2` is trivial, but `col1 + 1 as col2` is nontrivial.) If there are multiple matches, MySQL binds the reference to the leftmost match.
- The remaining selected columns, each of which is either a trivial column alias, or a non-aliased selected column. These have the same preference. If there are multiple matches, check whether they are actually the same underlying column. If so, use that. If not, return an ambiguous reference error.
- An unselected column from an underlying table. If there are multiple matches, return an ambiguous reference error.

Whereas `GROUP BY` uses the following order:

- First, check whether there are multiple matching columns that are either trivial column alias, or a non-aliased selected column. If so, check whether they are actually the same underlying column, and if not, return an ambiguous reference error. (This rule isn't actually used to bind anything, and I suspect it's also a bug.)
- A non-aliased selected column. If there are multiple matches, return an ambiguous reference error.
- An unselected column from an underlying table. If there are multiple matches, check whether they are actually the same underlying column. If so, use that. If not, return an ambiguous reference error.
- A nontrivial column alias. If there are multiple matches, MySQL binds the reference to the leftmost match.
- A trivial column alias. If there are multiple matches, check whether they are actually the same underlying column. If so, use that. If not, return an ambiguous reference error.

In addition, if either `GROUP BY` or `ORDER BY` uses an expression instead of a column name, MySQL uses a third resolution order. (This may also be a bug in MySQL.)

- A column from an underlying table, regardless of whether that column was selected. If there are multiple matches, return an ambiguous reference error.
- A nontrivial column alias. If there are multiple matches, MySQL binds the reference to the leftmost match.
- The remaining selected columns, each of which is either a trivial column alias, or a non-aliased selected column. These have the same preference. If there are multiple matches, check whether they are actually the same underlying column. If so, use that. If not, return an ambiguous reference error.

`sqllogictests` has tests specifically for the case of `GROUP BY` where the group expression is a single identifier (not a compound expression), and the query is expected to bind successfully. We need to support at least that case.

**Comment by max-hoffman** [comment]
Interesting! Very thorough investigative work. One thing I was confused about by the title is that I think we handle this in the absence of aggregates -- specifically `select col, col from tab group by col` and `select col, col from tab order by col` throw the ambiguous error.  There is a long tail of non-trivial aggregation validation, but I think this one should be pretty easy? `analyzeOrderBy` and `buildGroupingCols` are structured in a way to permit the specific precedence cases nick lists above.

**Comment by nicktobey** [comment]
That's good to hear that we already handle it in the absence of aggregates. That's half the work already done right there. Sounds like this is going to be a lot simpler to fix than I was fearing.

**Comment by max-hoffman** [comment]
Whoops I should have been more precise, the order by technically isn't an aggregate. Fixing ORDER BY is structurally the same as group by though. Windows might require the same fix:
```
select row_number() over () as col, row_number() over () as col from t order by col;
ambiguous column or alias name "i"
```
I meant I don't think this will require any of the more heroic changes some of the harder bugs need: unique ids for every expression, memoization, unnesting lateral subqueries, a post-planbuilder validator for aggregation functional dependency, ...
